src/oph/ehoks/opiskelijapalaute.clj

```diff
@@ -0,0 +1,128 @@
+             (not (:osaamisen-saavuttamisen-pvm hoks)))
+        (log/info msg "`osaamisen-saavuttamisen-pvm` has not yet been set "
+                  "for given HOKS.")
+        :else true)))
+  ([kysely current-hoks updated-hoks] ; on HOKS update
+    {:pre [(#{:aloituskysely :paattokysely} kysely)]}
+    (let [msg (format "Sending %s for HOKS `%s`. "
+                      (name kysely) (:id updated-hoks))]
+      (and (:osaamisen-hankkimisen-tarve updated-hoks)
+           (not (c/tuva-related-hoks? updated-hoks))
+           (if (= kysely :aloituskysely)
+             (not ; In case of log prints `nil` is returned, convert to `true`.
+               (cond
+                 (not (:osaamisen-hankkimisen-tarve current-hoks))
+                 (log/info msg "`osaamisen-hankkimisen-tarve` updated from "
+                           "`false` to `true`.")
+
+                 (added? :sahkoposti current-hoks updated-hoks)
+                 (log/info msg "`sahkoposti` has been added.")
+
+                 (added? :puhelinnumero current-hoks updated-hoks)
+                 (log/info msg "`puhelinnumero` has been added.")
+
+                 :else true)) ; will be converted to `false`.
+             (and
+               (added? :osaamisen-saavuttamisen-pvm current-hoks updated-hoks)
+               (not (log/info
+                      msg
+                      "`osaamisen-saavuttamisen-pvm` has been added."))))))))
+
+(defn send!
+  "Sends heräte data required for opiskelijapalautekysely (`:aloituskysely` or
+  `:paattokysely`) to appropriate DynamoDB table of Herätepalvelu."
+  [kysely hoks]
+  {:pre [(#{:aloituskysely :paattokysely} kysely)
+         (:id hoks)]}
+  (try
+    (case kysely
+      :aloituskysely
+      (sqs/send-amis-palaute-message (sqs/build-hoks-hyvaksytty-msg hoks))
+      ; In the case of päättökysely, `opiskeluoikeus` is fetched from Koski in
+      ; order to determine, if the kyselytyyppi is `tutkinnon_suorittaneet` or
+      ; `tutkinnon_osia_suorittaneet`."
+      :paattokysely
+      (or (some->> (:opiskeluoikeus-oid hoks)
+                   k/get-opiskeluoikeus-info
+                   get-kysely-type
+                   (sqs/build-hoks-osaaminen-saavutettu-msg hoks)
+                   sqs/send-amis-palaute-message)
+          (log/warnf (str "Not sending paattokysely for HOKS `%s`. Couldn't "
+                          "get opiskeluoikeus `%s` from Koski.`")
+                     (:id hoks)
+                     (:opiskeluoikeus-oid hoks))))
+    (catch Exception e
+      (log/warn e)
+      (log/warnf (str "Error in sending %s for HOKS `%s`. "
+                      "osaamisen-saavuttamisen-pvm %s."
+                      "opiskeluoikeus-oid %s.")
+                 (name kysely)
+                 (:id hoks)
+                 (:osaamisen-saavuttamisen-pvm hoks)
+                 (:opiskeluoikeus-oid hoks)))))
+
+(defn send-if-needed!
+  [kysely hoks]
+  (when (send? kysely hoks) (send! kysely hoks)))
```

Comment:
(toive) 
täs käytetään vähän salaa hyväksi sitä, että send! palauttaa `true`.  Eipä siinä mitään, mutta jos joku lukee tätä koodia, pitää hyppiä koodissa että näkee miksi tuo `filter` toimii, koska huutomerkillisistä funktioista ei voi nimestä päätellä, mitä ne palauttaa, eikä sitä oo myöskään dokumentoitu.  Ehkä pari docstringiä lisää vois tehdä hyvää.

Reply:
Nimittäin funktioihin `send!` ja `send-if-needed!`